### ISSUE TITLE
dateutil parser rounding or correctness problem?

### ISSUE THREAD

**Issue by dddmello** [dateutil/dateutil#427]
Was running some tests and the below code:

    from dateutil import parser as dparser
    print((dparser.parse('5.6h') - dparser.parse('0s')).total_seconds())

prints:

20100.0

I would expect it to print 20160.0 (off my 60 seconds).

Specifically it looks like:

>>> dparser.parse('5.6h')
datetime.datetime(2017, 7, 26, 5, 35)

Should actually be returning datetime.datetime(2017, 7, 26, 5, 36)

I see this behavior with version 2.5.3 - wondering if this is expected or if there is a rounding/truncation bug somewhere?

**Comment by jbrockmendel** [comment]
Looks like this is a floating point issue:

```
>>> 5.6 % 1
0.5999999999999996
>>> int(60*(5.6 % 1))
35
```

I'll put in a PR to fix this -- unless you want beat me to it.

**Comment by jbrockmendel** [comment]
> Was running some tests

If you have some real-life test cases we can add to the test suite, those would be helpful.

**Comment by pganssle** [comment]
I suppose the rounding issue should be fixed, though if this is indeed the issue, it seems to imply that "5.6h" only parses to hours + minutes, not hours + minutes + seconds + microseconds. I'm not really sure who is using "5.6h" as a time format rather than an interval format, but I'm somewhat surprised that to the extent we're supporting it, we truncate at minutes.

That said, this is not the right use of `dateutil.parser.parse`. "5.6h" and "0s" are intervals, not times, and unless you explicitly specify the `default`, there might be subtle bugs during certain transitions (e.g. if run at midnight), if you use them as such, since what's happening is you're essentially parsing "5.6h" into "today at 5AM + 0.6 hours" and "0s" as "today at midnight", then subtracting the two to get an actual interval. This is a *horrible* abuse of a a datetime parser.

I think the proper way to handle this would be to implement a separate interval parser that returns `timedelta` and/or `relativedelta` objects.l See #295 and #351. There are probably other issues that also boil down to "we should have an interval parser."

**Comment by jbrockmendel** [comment]
I've got a "fix" for this that parses float-like strings more carefully.  Should I bother making a PR or is this format going to be handled elsewhere?

**Comment by pganssle** [comment]
@jbrockmendel I think we can fix the rounding problem. This format is in the test suite so it's something we support. I'm just horrified by how it's being used here, so we should probably provide a proper alternative.

**Comment by m-dz** [comment]
@pganssle , I will give this one a try.

**Comment by pganssle** [comment]
:+1:

**Comment by m-dz** [comment]
@pganssle , two extra stupid use cases, what should they return (throw an error?):

```
# à la HH:MM[:SS[.ss]]
teststr = '05.6:00'
print(parse(teststr, default=datetime(2003, 9, 25, 0, 0)))
# Currently: print(datetime(2003, 9, 25, 5))
```

```
# Combined string with fractional h, m and s like 5.6h5.6m5.6s
# Should return:
# 2003-09-25 05:36:00 +
# 2003-09-25 00:05:36 +
# 2003-09-25 00:00:05.600000
# = 2003-09-25 05:41:41.600000
# or rather throw some sort of a parsing error?
teststr = '5.6h5.6m5.6s'
print(parse(teststr, default=datetime(2003, 9, 25, 0, 0)))
# Currently: print(datetime(2003, 9, 25, 5, 5, 5, 600000))
```

**Comment by pganssle** [comment]
@m-dz The first one I'm not sure why it's not picking up the 6:00 there, but it's hard to tell if it's just a manifestation of a well-known bug.

For the second one, I'm not even sure why we support `5.6s` or whatever *anyway*, it's not really a `datetime` format, but evidently someone is using it somewhere. I'm not interested in being prescriptive about what those sorts of formats *should* refer to. If there is evidence that they are in use, we should try to be descriptive about how they are actually used. If they are not in use (which I doubt they are), I'd prefer to raise an error - but I don't feel strongly enough about it to want to change anything.

**Comment by m-dz** [comment]
@pganssle Those two examples are completely artificial, I won't expect anyone using them. When fixing this issue I typed in a few weird formats to see if it breaks or not and it returned "incorrect" (?) values.

The `'05.6:00'` is split on ":" and parsed as `int(5.6)` for hours and `_parse_min_sec(value)` for minutes and seconds, which I would say is an expected behaviour. Maybe there should be some input validation to allow only 'HH[ ][h]:MM[ ][m]:SS[.ss][s]"?

The `'5.6h5.6m5.6s'` is tokenized into `['5.6', 'h', '5.6', 'm', '5.6' 's']` and then '5.6' is parsed into 5 hours 36 seconds, then the next '5.6' is parsed into 5 minutes 36 seconds, which overwrites minutes, then the last '5.6' is parsed into 5 seconds 600 ms, which overwrites seconds. Similarly, some input validation might help.

But all the above seems to be a completely unrelated issue to the rounding issue from this thread.

**Comment by pganssle** [comment]
Fixed with #570.
